### PR TITLE
Refactor uniswap, curve & interface, introduce extrinsic + benchmarkings

### PR DESCRIPTION
**Pull request by hussein-aitlahcen**

### CODE REVIEW

File: frame/curve-amm/src/lib.rs [hunk 1]
```diff
@@ -278,21 +275,104 @@ pub mod pallet {
...
+			let prec = T::Precision::get();
+			let assets_pair = PoolAssets::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+			let mut base_asset = assets_pair.base;
+			let mut quote_asset = assets_pair.quote;
+			if asset_id == quote_asset {
+				base_asset = assets_pair.quote;
+				quote_asset = assets_pair.base;
+			} else if asset_id != base_asset {
+				return Err(Error::<T>::AssetNotFound.into())
+			}
```
Comment (andor0) [inline frame/curve-amm/src/lib.rs:293]: imho, so easier to understand
```
let (base_asset, quote_asset) = match (assets_pair, asset_id) {
  (ap, id) if ap.base == id => (ap.base, ap.quote),
  (ap, id) if ap.quote == id => (ap.quote, ap.base),
  _ => return Err(Error::<T>::AssetNotFound.into())
};			};
```

File: frame/curve-amm/src/lib.rs [hunk 2]
```diff
@@ -278,21 +275,104 @@ pub mod pallet {
...
+			// -1 just in case there were some rounding errors
+			// dy = xp[j] - y - 1
+			let xp_j = xp.get(1).ok_or(Error::<T>::InconsistentStorage)?;
+			let dy_f = xp_j
+				.checked_sub(&y)
+				.ok_or(Error::<T>::Math)?
+				.checked_sub(&prec)
+				.ok_or(Error::<T>::Math)?;
+
+			// let dy: Self::Balance = dy_f.checked_mul_int(1_u64).ok_or(Error::<T>::Math)?.into();
```
Comment (andor0) [inline frame/curve-amm/src/lib.rs:322]: Do you forget to remove it?

File: frame/curve-amm/src/lib.rs [hunk 3]
```diff
@@ -601,172 +673,114 @@ pub mod pallet {
...
-			for asset_id in assets {
-				balances.push(PoolAssetBalance::<T>::get(pool_id, asset_id));
+			let mut base_asset = assets_pair.base;
+			let mut quote_asset = assets_pair.quote;
+			if asset_id == quote_asset {
+				base_asset = assets_pair.quote;
+				quote_asset = assets_pair.base;
+			} else if asset_id != base_asset {
+				return Err(Error::<T>::AssetNotFound.into())
 			}
```
Comment (andor0) [inline frame/curve-amm/src/lib.rs:692]: This part of code has been used a minimum of 2 times, maybe you should create a function for it?

**Comment by dzmitry-lahoda** [comment]
can we merge working DEX ? just enough not to brick parachain, but may be not so good in terms of overflows, but working?

**Comment by hussein-aitlahcen** [comment]
> can we merge working DEX ? just enough not to brick parachain, but may be not so good in terms of overflows, but working?

There is 2 branches that have work that will get merged into this one, it's almost done.

**Review by benluelo — DISMISSED** [review]

**Review by KaiserKarel — APPROVED** [review]

**Comment by hussein-aitlahcen** [comment]
Awaiting for #684 to trigger CI benchmarking generation.

**Review by benluelo — APPROVED** [review]

**Review by PoisonPhang — APPROVED** [review]

**Pull request merged**